docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+
+Imagine a cluster that is running pods from two services, S1 and S2. Imgine that the pods of S1 have a hard anti-affinity rule
+"do not run me on nodes that are running pods from S2." It is not sufficient just to check that there are no S2 pods on a node when
+you are scheduling a S1 pod. You also need to ensure that there are no S1 pods on a node when you are scheduling a S2 pod,
+*even though the S2 pod does not have any anti-affinity rules*. Otherwise if an S1 pod schedules before an S2 pod, the S1
+pod's hard anti-affinity rule can be violated by a later-arriving S2 pod.
+
+Note that while hard anti-affinity is symmetric,
+hard affinity is *not* symmetric. That is, if the pods of S1 have a hard affinity rule "run me on nodes that are running
+pods from S2," it is not required that there be S1 pods on a node in order to schedule a S2 pod onto that node.
+
+Soft affinity and soft anti-affinity are both symmetric.
+If the pods of S1 had a soft anti-affinity rule "try not to run me on nodes that are running pods from S2"
+then we would prefer to keep a S1 pod that we are scheduling off of nodes that are running S2 pods, and also
+to keep a S2 pod that we are scheduling off of nodes that are running S1 pods. Likewise if the pods of
+S1 had a soft affinity rule "try to run me on nodes that are running pods from S2" then we would prefer
+to place a S1 pod that we are scheduling onto a node that is running a S2 pod, and also to place
+a S2 pod that we are scheduling onto a node that is running a S1 pod.
+
+As one further twist, although
+hard affinity is not symmetric, there is an implicit soft affinity rule corresponding to every
+hard affinity rule: if the pods of S1 have a hard affinity rule "run me on nodes that are running
+pods from S2" then it is not required that there be S1 pods on a node in order to schedule a S2 pod onto that node,
+but it would be better if there are.
+
+## Examples
+
+Here are some examples of how you would express various affinity and anti-affinity rules using the API we described.
+
+### Affinity
+
+In the examples below, the word "put" is intentionally ambiguous; the rules are the same
+whether "put" means "must put" (hard) or "try to put" (soft)--all that changes is whether
+the rule goes in HardAffinityExpressions or SoftAffinityExpressions. Also, some of the examples
+use "zone" and some use "node," just to make the examples more interesting; any of the examples
+with "zone" will also work for "node" if you change the `TopologyKey`, and vice-versa.
+
+* **Put the pod in zone Z**:
+Tricked you! It is not possible express this using the API described here. For this you should use a NodeConstraint.
+
+* **Put the pod in a zone that is running at least one pod from service S**:
+`{PodSelector: <selector that matches S's pods>, TopologyKey: "zone"}`
+
+* **Put the pod on a node that is already running a pod that requires a license for software package P**:
```

Comment:
Ok, ignore my comment above about software licensing.  I guess you're targeting the "per client machine" license model here, which Oracle and others have been known to use.  It would be worth working through some of the details here, as I'm not sure that this will actually work correctly in practise.  For example, many licensing schemes use client IP address, and if we don't preserve that correctly, that won't work.  Others use some sort of cert to install onto each machine, but them means exposing that cert to all of the client containers/Pods on the machine, and I'm not sure we've thought through that.